engine/crates/engine/src/registry/resolvers/postgresql/context/selection/collection_args.rs

```diff
@@ -0,0 +1,161 @@
+use crate::{registry::resolvers::postgresql::context::PostgresContext, SelectionField};
+use engine_value::{Name, Value};
+use grafbase_sql_ast::ast::{Aliasable, Column, Order, OrderDefinition};
+use indexmap::IndexMap;
+use postgresql_types::database_definition::TableWalker;
+use std::borrow::Cow;
+
+/// Argument defining a relay-style GraphQL collection.
+#[derive(Debug, Clone)]
+pub struct CollectionArgs {
+    first: Option<u64>,
+    last: Option<u64>,
+    order_by: Option<(Vec<OrderDefinition<'static>>, Vec<OrderDefinition<'static>>)>,
+    extra_columns: Vec<Column<'static>>,
+    before: Option<String>,
+    after: Option<String>,
+}
+
+impl CollectionArgs {
+    pub(super) fn new(ctx: &PostgresContext<'_>, table: TableWalker<'_>, value: &SelectionField<'_>) -> Self {
+        let first = value.field.get_argument("first").and_then(|value| value.as_u64());
+        let last = value.field.get_argument("last").and_then(|value| value.as_u64());
+        let before = value.field.get_argument("before").and_then(|value| value.as_string());
+        let after = value.field.get_argument("after").and_then(|value| value.as_string());
+
+        let mut order_by_argument = value
+            .field
+            .get_argument("orderBy")
+            .and_then(|value| value.as_slice())
+            .map(Cow::from);
+
+        // For `last` to work, we need some known order for the inner query.
```

Comment:
why is `first` any different? you have the same problem in both directions with cursor-based pagination. with `first: 100, after: X` we need to be sure `X` is unique.

Reply:
This is more of a notion that we do `LIMIT` here too, but the order of the inner query is different.

Reply:
not sure what you mean, you're only specifying an ordering with `last`, not with `first`. Even though `first` still needs some ordering due to cursor navigation. So I don't get that part.

Reply:
Hmm, maybe I should enforce ordering then. I was thinking that it comes implicitly from the clustered primary or if missing, secondary key. Maybe I should not rely on that...

Reply:
no idea, would need to check postgres doc for several versions at least. It's probably the case. My point is mostly than `first` and `last` aren't different in that regard, so if you need to add ordering for one, the other one isn't any different.